peer/peer.go

```diff
@@ -0,0 +1,2114 @@
+func (p *Peer) LastSend() time.Time {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.lastSend
+}
+
+// LastRecv returns the last recv time of the peer.
+func (p *Peer) LastRecv() time.Time {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.lastRecv
+}
+
+// BytesSent returns the bytes sent by the peer.
+func (p *Peer) BytesSent() uint64 {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.bytesSent
+}
+
+// BytesReceived returns the bytes received by the peer.
+func (p *Peer) BytesReceived() uint64 {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.bytesReceived
+}
+
+// TimeConnected returns the time at which the peer connected.
+func (p *Peer) TimeConnected() time.Time {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.timeConnected
+}
+
+// TimeOffset returns the time offset from the peer.
+func (p *Peer) TimeOffset() int64 {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.timeOffset
+}
+
+// StartingHeight returns the starting height of the peer.
+func (p *Peer) StartingHeight() int32 {
+	p.StatsMtx.Lock()
+	defer p.StatsMtx.Unlock()
+
+	return p.startingHeight
+}
+
+// pushVersionMsg sends a version message to the connected peer using the
+// current state.
+func (p *Peer) pushVersionMsg() error {
+	_, blockNum, err := p.NewestSha()
+	if err != nil {
+		return err
+	}
+
+	theirNa := p.na
+
+	// If we are behind a proxy and the connection comes from the proxy then
+	// we return an unroutable address as their address. This is to prevent
+	// leaking the tor proxy address.
+	if p.cfg.Proxy != "" {
+		proxyaddress, _, err := net.SplitHostPort(p.cfg.Proxy)
+		// invalid proxy means poorly configured, be on the safe side.
+		if err != nil || p.na.IP.String() == proxyaddress {
+			theirNa = &wire.NetAddress{
+				Timestamp: time.Now(),
+				IP:        net.IP([]byte{0, 0, 0, 0}),
+			}
+		}
+	}
+
+	// Version message.
+	msg := wire.NewMsgVersion(
+		p.addrManager.GetBestLocalAddress(p.na), theirNa,
+		p.nonce, int32(blockNum))
+	msg.AddUserAgent(p.cfg.UserAgentName, p.cfg.UserAgentVersion)
+
+	// XXX: bitcoind appears to always enable the full node services flag
+	// of the remote peer netaddress field in the version message regardless
+	// of whether it knows it supports it or not.  Also, bitcoind sets
+	// the services field of the local peer to 0 regardless of support.
+	//
+	// Realistically, this should be set as follows:
+	// - For outgoing connections:
+	//    - Set the local netaddress services to what the local peer
+	//      actually supports
+	//    - Set the remote netaddress services to 0 to indicate no services
+	//      as they are still unknown
+	// - For incoming connections:
+	//    - Set the local netaddress services to what the local peer
+	//      actually supports
+	//    - Set the remote netaddress services to the what was advertised by
+	//      by the remote peer in its version message
+	msg.AddrYou.Services = p.Services()
```

Comment:
This is incorrect.  The remote peer services is not the same as the local peer.  It should be set to `SFNodeNetwork` as it was before.